feat: entity decorator

jrista opened the pull request:
+ Add @Entity decorator for models
 + Implement required modelName option
 + Implement optional pluralName option
 + Implement optional uriName option
 + Implement optional comparer function option
 + Implement effect exclusion feature
 + Add sorted$ facade stream and selectAllSorted selector
 * Fix prod build issue with modelName in @Entity decorator
 * Improve robustness of all selectors when state is falsy

 + Added test-app for simpler app than demo to test library features with

Issue #70 #58 #81

patpaquette commented on projects/ngrx-auto-entity/src/lib/util.ts:
```diff
@@ -541,12 +576,29 @@ export const buildFeatureState = <TState extends IEntityState<TModel>, TParentSt
+  const instance = new type();
+  console.log(`NGRX-AE: Building entity for: ${type.name}; constructor: ${instance.constructor.name}`);
+  const opts = type[ENTITY_OPTS_PROP] || {
+    modelName: instance.constructor.name
```
Will this be a problem while this bug is open? https://github.com/webpack-contrib/uglifyjs-webpack-plugin/issues/269

jrista replied:
This is actually why I made these changes as I came across that issue while researching the problems both mxdmedia and I have had. Previously, I was only using `instance.constructor.name`. What I've done is add the ability to override the name of the model using  the new @Entity decorator, which attaches the ENTITY_OPTS_PROP to the type itself. That way, when the constructor.name gets mangled, your explicitly specified name will still work. This has been tested with prod builds and it works. I kept the constructor.name here as a fallback for now.

patpaquette replied:
Gotcha, makes sense.

Commit:
feat: entity decorator

 + Add @Entity decorator for models
 + Implement required modelName option
 + Implement optional pluralName option
 + Implement optional uriName option
 + Implement optional comparer function option
 + Implement effect exclusion feature
 + Add sorted$ facade stream and selectAllSorted selector
 * Fix prod build issue with modelName in @Entity decorator
 * Improve robustness of all selectors when state is falsy

 + Added test-app for simpler app than demo to test library features with

Issue #70 #58 #81
